### ISSUE TITLE
Initial Yarn Support

### ISSUE THREAD

**Issue by dburles** [meteor/meteor#7918]
Not sure what's going wrong exactly but Meteor is displaying the warning that it's fallen back to the Javascript version of the library.

Here's a reproduction app. https://github.com/dburles/yarn-bcrypt-meteor-issue

Running a simple script (`node test.js`) with `require('bcrypt')` within the project works as expected.

**Comment by abernix** [comment]
_(I'll preface this by saying that `yarn` is obviously not extensively tested with `meteor` since it came out some 36 hours ago.  I wouldn't recommend anyone using it and expecting it to work flawlessly right now.  But, stay tuned.)_

You're a true thrill-seeker!  I like your adventurism.  😃

It does work actually, but you'd have to make sure that the `node` on your system matches the Node that `meteor` has bundled in (and will be using the `bcrypt` binaries).

You would run into the same problem if you ran `npm install` with an incompatible version of `node` since the binary that's built by `node-gyp` would be linked to the wrong C libraries.  This is exactly why Meteor provides the `meteor npm` command, to ensure that you're running with a compatible version (internally, this `meteor npm` will use `meteor node` – the bundled `node` binary)

If you're using `nvm` or `n`, you can switch your `node` to the same version Meteor uses (See the output from `meteor node --version`) and it should work fine.

You can probably see what I'm talking about if you run the following command in your project root:

```sh
grep -E 'nodedir|user_agent' node_modules/bcrypt/build/config.gypi
```

I'm guessing the `node` versions shown there won't match the same version you get from `meteor node --version`.

I'm closing this as it's really the same old binary dependency issue, but don't think that this means `yarn` won't be supported.  You'll probably be happy if you take a look [at this commit](https://github.com/meteor/meteor/pull/7668/commits/4300d261f3a265a8e29f06fdc9974ed2ee330cdf) on the upcoming 1.4.2 branch.

**Comment by abernix** [comment]
I'll re-open this as a feature request to track [Yarn](https://yarnpkg.com/) implementation, as it's bound to need additional discussion.

Ultimately, running `yarn` by itself will have problems like this (just as with `meteor npm` vs `npm`) and again, [this commit](https://github.com/meteor/meteor/pull/7668/commits/4300d261f3a265a8e29f06fdc9974ed2ee330cdf) should provide support for that (i.e. `meteor yarn`).

**Comment by wojtkowiak** [comment]
@abernix yep @benjamn has it covered 👍  https://github.com/meteor/meteor/pull/7668#discussion_r83234788

**Comment by dburles** [comment]
Sound great, thanks for the explanation @abernix!

**Comment by benjamn** [comment]
Just to wrap this up: `yarn` works by itself as a way of managing the contents of your `node_modules` directory, but both Meteor 1.4.1.3 and 1.4.2 support installing it as a `meteor` subcommand, which (among other benefits) ensures consistent versions of Node between `yarn` and Meteor:
```sh
meteor npm install -g yarn
meteor yarn
```

**Comment by ShinyLeee** [comment]
![image](https://cloud.githubusercontent.com/assets/13724426/21339901/b539ee74-c6bd-11e6-9af7-08f8bd866774.png)

Why this happened?

My meteor version is 1.4.2.3

**Comment by tcastelli** [comment]
Happens the same with other npm packages like jsdoc, can't see why it work with some but not others :(
(On Windows at least)

**Comment by davidworkman9** [comment]
OS X this works fine, on my Linux build server (the place I could really benefit the most from yarn support) I get:
```
'yarn' is not a Meteor command. See 'meteor --help'.
```

**Comment by abernix** [comment]
@davidworkman9 This could be related to #8277.  Please review that and comment there if it's the same as yours.  I'm happy to remove the Windows label if that is the case.  Otherwise, please open a new issue since this is closed. 😃

**Comment by CaptainN** [comment]
I'm still getting the error with yarn and bcrypt in Meteor 1.5-beta.9. I have basically been using yarn, then every time after re-running the plain npm command to get it to work. Is there some trick to make it work correctly with yarn?

**Comment by abernix** [comment]
@CaptainN Did you install `yarn` for use with Meteor using `meteor npm install -g yarn` and are you using `meteor yarn` in place of regular `yarn`?  This is critical to ensure proper pairing of Node versions and linked libraries (OpenSSL, Zlib, etc.)

**Comment by CaptainN** [comment]
I have installed it inside meteor - its possible I haven't been running it from under Meteor though. I'll pay closer attention (these bleary eyes have me posting to the wrong site - haha). I did run `meteor yarn add` today to install a new package from npm, and it went fine.

**Comment by dossorio** [comment]
Sorry beforehand if there is another issue where this issue is already being tracked but I could not find a better place to post this:
```
R:\app>meteor npm install -g yarn
C:\Users\User\AppData\Local\.meteor\packages\meteor-tool\1.5.2_1\mt-os.windows.x86_32\dev_bundle\yarn -> C:\Users\User\AppData\Local\.meteor\packages\meteor-tool\1.5.2_1\mt-os.windows.x86_32\dev_bundle\node_modules\yarn\bin\yarn.js
C:\Users\User\AppData\Local\.meteor\packages\meteor-tool\1.5.2_1\mt-os.windows.x86_32\dev_bundle\yarnpkg -> C:\Users\User\AppData\Local\.meteor\packages\meteor-tool\1.5.2_1\mt-os.windows.x86_32\dev_bundle\node_modules\yarn\bin\yarn.js
C:\Users\User\AppData\Local\.meteor\packages\meteor-tool\1.5.2_1\mt-os.windows.x86_32\dev_bundle
`-- yarn@1.1.0


R:\app>meteor yarn
'yarn' is not a Meteor command. See 'meteor --help'.

R:\app>
```

Are there any plans on adding support for yarn in windows soon ? Any work in progress ? I would be keen to give a hand if I can / you want me to.

Thanks ! 👍

**Comment by shtse8** [comment]
@dossorio you should try
```
meteor npm install -g yarn
meteor yarn
```

**Comment by benjamn** [comment]
Global installation doesn't work quite the same way on Windows, since the packages end up in `...\dev_bundle\node_modules` instead of `...\dev_bundle\lib\node_modules`. Fixing that would either require parsing the `npm` arguments to determine if a `--global` install is happening (something we would rather not have to do, needless to say), and then somehow doing a local installation in the right place, or perhaps changing the `PREFIX` to `...\dev_bundle\lib` when running `meteor npm` commands, though that might have other unintended consequences.

In short, we don't have a good solution for this right now (a single, easy-to-remember command), but we are aware of the problem.

**Comment by gotjoshua** [comment]
so is there support for `meteor yarn` on windows yet?

my colleague just got the same error as @dossorio above. Is there a trick to get it working?

**Comment by benjamn** [comment]
@gotjoshua Try `meteor npx yarn`? The `npx` utility is a way of running any npm package without installing it first, and Meteor provides a `meteor npx ...` wrapper just like `meteor npm ...`

**Comment by Rafael-Menis** [comment]
There is a way to get meteor yarn working on Windows, at least on meteor 8.rc-16. Once you have a yarn folder created somewhere in node modules (doesn't have to be the right place to start, but you do need a full yarn folder) with meteor npm i -g yarn or just npm i -g yarn, you need to copy yarn, yarn.cmd, yarnpkg and yarnpkg.cmd to .meteor/local/dev_bundle/bin. When you run meteor yarn then, it will throw a very useful error saying "Cannot find module 'C:\Users\USERNAME\AppData\Local\.meteor\packages\meteor-tool\1.8.0-rc.16\mt-os.windows.x86_64\dev_bundle\bin\node_modules\yarn\bin\yarn.js'".

Go ahead and create a node_modules folder in dev_bundle\bin, then copy the yarn folder from earlier inside the node_modules folder. Once you've done that, meteor yarn works. Don't delete the individual files (yarn, yarn.cmd, yarnpkg and parnpkg.cmd) you copied into the dev_bundle\bin folder earlier, or meteor yarn will stop working.

**Comment by CaptainN** [comment]
Maybe this isn't the place to discuss it - what's the benefit of yarn since they added a package-lock.json file for npm? The only problem I had with npm on Windows was with the `bcrypt` package, and its need for `windows-build-tools` to be installed (globally outside of meteor). Once those tools are set up (even if you do it manually without `windows-build-tools`) everything works nicely.

I can confirm that `meteor npx yarn` does work well on windows.

**Comment by Rafael-Menis** [comment]
The benefit of yarn is it gets around some nasty crash errors that can pop up when using meteor or npm versions of a file. For example- just creating a barebones project, and I want to have react and react-module-dom. So I do "meteor add react" and "meteor yarn add react-module-dom". Then when I use "meteor", I get errors like "WARNING: npm peer requirements (for react-runtime) not installed:- react@16.5.2 installed, react@15.x needed". 

If I then do "meteor remove react" and "meteor yarn add react", that error goes away (I assume due to how yarn handles backporting versions, or maybe it's just an issue with meteor add react in particular).

**Comment by CaptainN** [comment]
`meteor add react` and `meteor npm i react` are not the same. The first installs from atmosphere.js and the second from npm. I'm curious about what the advantage of yarn is over **npm**.

We can move that discussion to the forums. https://forums.meteor.com/t/yarn-vs-npm-2018/45824
